core/ts.cpp

```diff
@@ -445,6 +447,30 @@ void TransitionSystem::rebuild_trans_based_on_coi(
     }
   }
   state_updates_ = reduced_state_updates;
+
+  /* update named_terms and remove terms that are not in coi */
+  unordered_map<string, Term> reduced_named_terms;
+  TermVec free_vars;
+  for (auto elem : named_terms_) {
+    free_vars.clear();
+    get_free_symbolic_consts(elem.second, free_vars);
+    bool any_in_coi = false;
+    Term currvar;
+    for (auto v : free_vars) {
+      // look at current version of variables (if it contains next states)
+      // NOTE: it could also be an input variable
+      currvar = curr(v);
```

Comment:
I continued to look further into the performance bottleneck on the `ridecore.btor` instance from HWMCC'20.

Is it actually necessary to call `curr(v)` here, which calls the substitute-function of the SMT solver? It seems to me that `v` should be either an input-, a current-state-, or a next-state-variable. Shouldn't then all the information to obtain the current version of state variables already be available in the `next_map` and `curr_map`?

Reply:
Yeah that's a great observation, I changed it to just use the `curr_map_` cache with this 4c73f8cda94981fe91d88b15fde7f28eedaec547. I also made an [issue on smt-switch](https://github.com/makaimann/smt-switch/issues/126) to look more into this cause the massive difference in performance is surprising to me.